File: tests/src/org.joda/joda-convert/2.2.4/src/test/java/org_joda/joda_convert/TypesInnerTypeVariableInvocationHandlerTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_joda.joda_convert;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.lang.reflect.GenericDeclaration;
import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.lang.reflect.TypeVariable;

import org.joda.convert.StringConvert;
import org.junit.jupiter.api.Test;

public class TypesInnerTypeVariableInvocationHandlerTest {
    @Test
    public void artificialTypeVariablesDelegateSupportedMethodsToBackingImplementation() throws Exception {
        TypeVariable<?> variable = newArtificialTypeVariable(
                TypesInnerTypeVariableInvocationHandlerTest.class,
                "Synthetic",
                Number.class);

        assertEquals("Synthetic", variable.getName());
        assertSame(TypesInnerTypeVariableInvocationHandlerTest.class, variable.getGenericDeclaration());
        assertEquals("Synthetic", variable.toString());
    }

    @Test
    public void handlesParameterizedTypesWithWildcardBounds() {
        StringConvert convert = StringConvert.create();
        ParameterizedType type = convert.convertFromString(
                ParameterizedType.class,
                "java.util.List<? super java.lang.Integer>");

        assertTrue(convert.convertToString(ParameterizedType.class, type).contains("java.lang.Integer"));
    }

    private static TypeVariable<?> newArtificialTypeVariable(
            GenericDeclaration declaration, String name, Type... bounds) throws Exception {
        Method newArtificialTypeVariable = Class.forName("org.joda.convert.Types")
                .getDeclaredMethod(
                        "newArtificialTypeVariable",
                        GenericDeclaration.class,
                        String.class,
                        Type[].class);
        newArtificialTypeVariable.setAccessible(true);
        return (TypeVariable<?>) newArtificialTypeVariable.invoke(null, declaration, name, bounds);
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/TypesInnerTypeVariableInvocationHandlerTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.lang.reflect.GenericDeclaration
import java.lang.reflect.Method
import java.lang.reflect.Type
import java.lang.reflect.TypeVariable

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test

class TypesInnerTypeVariableInvocationHandlerTest {
  @Test
  def artificialTypeVariablesDelegateSupportedMethodsToBackingImplementation(): Unit = {
    val variable: TypeVariable[?] = newArtificialTypeVariable(
      classOf[TypesInnerTypeVariableInvocationHandlerTest],
      "Synthetic",
      classOf[Number]
    )

    assertEquals("Synthetic", variable.getName)
    assertSame(classOf[TypesInnerTypeVariableInvocationHandlerTest], variable.getGenericDeclaration)
    assertEquals("Synthetic", variable.toString)
  }

  private def newArtificialTypeVariable(
      declaration: GenericDeclaration,
      name: String,
      bounds: Type*
  ): TypeVariable[?] = {
    val newArtificialTypeVariable: Method = Class
      .forName("org.joda.convert.Types")
      .getDeclaredMethod(
        "newArtificialTypeVariable",
        classOf[GenericDeclaration],
        classOf[String],
        classOf[Array[Type]]
      )
    newArtificialTypeVariable.setAccessible(true)

    newArtificialTypeVariable
      .invoke(
        null,
        declaration.asInstanceOf[Object],
        name,
        bounds.toArray.asInstanceOf[Object]
      )
      .asInstanceOf[TypeVariable[?]]
  }
}
